[IDE] Implement completion-like cursor info for ValueDecls

ahoppen opened the pull request:
This brings up the ability to compute cursor info results using the completion-like type checking paradigm, which an reuse ASTContexts and doesn’t need to type check the entire file.

For now, the new implementation only supports cursor info on `ValueDecl`s (not on references) because they were easiest to implement. More cursor info kinds are coming soon.

At the moment, we only run the new implementation in a verification mode: It is only invoked in assert toolchains and when run, we check that the results are equivalent to the old implementation. Once more cursor info kinds are implemented and if the SourceKit stress tester doesn’t find any verification issues, we can enable the new implementation, falling back to the old implementation if the new one didn’t produce any results.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen committed:
[IDE] Implement completion-like cursor info for ValueDecls

This brings up the ability to compute cursor info results using the completion-like type checking paradigm, which an reuse ASTContexts and doesn’t need to type check the entire file.

For now, the new implementation only supports cursor info on `ValueDecl`s (not on references) because they were easiest to implement. More cursor info kinds are coming soon.

At the moment, we only run the new implementation in a verification mode: It is only invoked in assert toolchains and when run, we check that the results are equivalent to the old implementation. Once more cursor info kinds are implemented and if the SourceKit stress tester doesn’t find any verification issues, we can enable the new implementation, falling back to the old implementation if the new one didn’t produce any results.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test macOS

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen committed:
[IDE] Run RecontextualizeClosures in TypeCheckASTNodeAtLocRequest

ahoppen committed:
[IDE] Decrement the global closure discriminator by one before performing the second code completion pass

Otherwise, the closure discriminator will be incremented by one when the closure witht he code completion token is parsed a second time during the second pass, and thus it would receive a different discriminator during the second pass.

ahoppen committed:
[SourceKit] Don't ignore swiftsourceinfo for solver-based cursor info

Cursor infor needs source info for referenced decls.

ahoppen commented:
@swift-ci Please smoke test macOS

ahoppen commented:
@swift-ci Please SourceKit stress test

rintaro commented on test/SourceKit/CursorInfo/cursor_in_global_closure.swift:
```diff
@@ -0,0 +1,6 @@
+fileprivate let formatter: DateFormatter = {
+    let formatter
+
+// CHECK: s:24cursor_in_global_closureXefU_9formatterL_Xevp
```
newline at EOF

rintaro commented on test/SourceKit/CursorInfo/cursor_uses_swiftsourceinfo.swift:
```diff
@@ -0,0 +1,20 @@
+// CHECK: REFERENCED DECLS BEGIN
+// CHECK-NEXT: s:8MyModule6ActionP | public | {{.*}}/split/Action.swift | MyModule | User | NonSPI | source.lang.swift
+// CHECK-NEXT: Action swift.protocol s:8MyModule6ActionP
+// CHECK-NEXT: REFERENCED DECLS END
```
newline

ahoppen commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -1906,9 +1932,130 @@ void SwiftLangSupport::getCursorInfo(
+  // result.
+  auto ReceiverThunk = [Receiver, VerifySolverBasedResult](
+                           const RequestResult<CursorInfoData> &Res) {
+#ifndef NDEBUG
```
Actually, I think that check is not necessary at all because `VerifySolverBasedResult` will just return if `SolverBasedResultDescription` is empty. I’ll just remove it.

ahoppen commented on test/SourceKit/CursorInfo/completion_like_cursor_with_edit.swift:
```diff
@@ -0,0 +1,10 @@
+// RUN: %sourcekitd-test -req=open %s -- %s == \
+// RUN: -req=edit -pos=7:16 -length=0 -replace=windowScenewindowScene %s == \
+// RUN: -req=cursor -pos=7:16 %s -- %s 
```
I had a problem where I wasn’t respecting edits to the file, so I thought I’d add a test for it. I can add a comment explaining that.

ahoppen replied:
Oh, I didn’t realize I forgot to add a `FileCheck` here. This was crashing in the verification before.

ahoppen commented on lib/IDE/CursorInfo.cpp:
```diff
@@ -0,0 +1,278 @@
+    return Action::Continue();
+  }
+
+  PreWalkResult<Expr *> walkToExprPre(Expr *E) override {
```
Is it OK if I do that in a follow-up PR. I’m not entirely sure if that causes new issues and I’d like to merge this one.

ahoppen commented on lib/IDE/CursorInfo.cpp:
```diff
@@ -0,0 +1,278 @@
+
+/// Walks the AST, looking for a node at \c LocToResolves. While walking the
+/// AST, also gathers information about shorthand shadows.
+class NodeFinder : ASTWalker {
```
I’ll see what I can do in a follow-up PR.

bnbarham replied:
Yeah, IMO ideally we'd do what Clang does here - ie. build up a hierarchy to the deepest node and pass that around everywhere we need it (rather than continually performing the same ASTWalk skipping outside-of-loc ranges in basically every refactoring + cursor info).

ahoppen replied:
I filed rdar://103178329 so we don’t forget about it and hope that I’ll get to it at some point.

bnbarham commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -1906,9 +1932,130 @@ void SwiftLangSupport::getCursorInfo(
+        }
+        // The solver-based implementation differed from the AST-based
+        // implementation. Report a failure.
+        llvm::errs() << SolverBasedResultDescription << "\n";
+        llvm::errs() << "===== (solver based vs legacy) =====\n";
+        llvm::errs() << ASTResultDescription << "\n";
+        assert(false);
```
Any thoughts on returning an error result with this text rather than crashing?

ahoppen replied:
Good idea 👍

bnbarham commented on lib/IDETool/CompletionInstance.cpp:
```diff
@@ -794,3 +794,70 @@ void swift::ide::CompletionInstance::conformingMethodList(
       });
 }
+
+void swift::ide::CompletionInstance::cursorInfo(
```
Little weird it's a part of `CompletionInstance` 😅.

ahoppen replied:
Yes, we should rename `CompletionInstance`. But I didn’t know how else to name it.

bnbarham commented on lib/IDE/CursorInfo.cpp:
```diff
@@ -0,0 +1,278 @@
+
+// MARK: - NodeFinder
+
+/// Walks the AST, looking for a node at \c LocToResolves. While walking the
```
```suggestion
/// Walks the AST, looking for a node at \c LocToResolve. While walking the
```

bnbarham commented on test/SourceKit/CursorInfo/cursor_in_pound_if.swift:
```diff
@@ -0,0 +1,13 @@
+  let xxx = "hello"
+#endif
+}
+// TODO: Once we switch to use the solver-based cursor info implementation, we also receive results for the int case
```
Does it handle unparsed `#if` regions?

ahoppen replied:
Yes, it just fell out of the implementation that it does. I think we just type check whatever decl the cursor info points to. I’m not entirely sure how well it works if the decl you invoke cursor info on references other variables in a `#if` though.

bnbarham commented on test/SourceKit/CursorInfo/cursor_uses_swiftsourceinfo.swift:
```diff
@@ -0,0 +1,20 @@
+// RUN: %{python} %utils/split_file.py -o %t/split %s
+// RUN: %empty-directory(%t/build)
+// RUN: %target-swift-frontend -emit-module -module-name MyModule -emit-module-path %t/build/MyModule.swiftmodule -emit-module-source-info-path %t/build/MyModule.swiftsourceinfo %t/split/Action.swift
+// RUN: %sourcekitd-test -req=cursor -req-opts=retrieve_symbol_graph=1 -pos=5:14 %t/split/test.swift -- %t/split/test.swift -I %t/build | %FileCheck %s
```
Any reason to include the symbol graph here rather than just doing cursor info on `Action` and checking the file/line/col? I would have thought we have a check for that already though, do we really not?

bnbarham replied:
Doing a cursor info on `Action` should also end up using the `.swiftsourceinfo`. `REFERENCED DECLS` does as well because the decl is actually the type of the decl you're doing cursor info on. But as I said, I'm very surprised we don't already have a test for that.

ahoppen replied:
For reasons I don’t fully understand (I haven’t really done any investigation though) we report the correct source location of `Action` even when we’re not using `.swiftsourceinfo`.

bnbarham commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -1906,9 +1932,130 @@ void SwiftLangSupport::getCursorInfo(
+    } else {
+      std::string Description;
+      llvm::raw_string_ostream OS(Description);
+      // The ContainerTypeUSR varies too much between the solver-based and
+      // AST-based implementation. A few manual inspections showed that the
+      // solver-based container is usually more correct than the old. Instead of
+      // fixing the AST-based container type computation, exclude the container
+      // type from the verification.
```
Where is the container type USR being filtered out?

ahoppen replied:
In the next PR (not open yet). I removed the comment for now.

ahoppen commented:
Note to self: The stress tester failure at offset 300 of modification `concurrent-359` of `rows/MovieRow.swift` has the following reduce reproducer:

```swift
// RUN: %sourcekitd-test -req=cursor -req-opts=retrieve_symbol_graph=1 -pos=5:12 %s -- %s

let formatter: DateFormatter = {
  struct MovieRow {
    struct Props
```

The solver-based implementation reports `Props` as conforming to `Sendable` in symbol graph, the AST-based implementation does not report this. I think this change is acceptable and I’ll just mark this discrepancy as an XFail in the stress tester.

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test macOS

ahoppen committed:
[IDE] Disable a few code completion specific paths for solver-based CursorInfo

The lexer will be responsible for knowing whether we have a code completion token, everything else will also work for other IDE inspection features.

The changes start to really make sense once I rename CodeCompletion -> IDEInspection in a lot of places.

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test
